cdf_fabric_replicator/extractor.py

```diff
@@ -194,51 +195,61 @@ def upload_files_to_cdf(
+            "https://storage.azure.com/.default"
+        ).token
+        self.logger.debug(f"Extracting time series data from {file_path}")
+        external_ids = self.retrieve_external_ids_from_lakehouse(file_path, token=token)
+        self.logger.debug(f"External IDs found: {external_ids}")
+        latest_timestamps = self.get_timeseries_latest_timestamps(external_ids)
+        for external_id, timestamp in latest_timestamps.items():
+            for time_series_data in self.convert_lakehouse_data_to_df_batch(
+                file_path,
+                external_id,
+                timestamp,
+                token=token,
+            ):
+                self.write_time_series_to_cdf(external_id, time_series_data)
+
+    def get_timeseries_latest_timestamps(self, external_ids: list[str]) -> dict:
+        latest_timestamps = {}
         for external_id in external_ids:
-            df = data_frame[data_frame["externalId"] == external_id]
             state_id = f"{self.config.source.raw_time_series_path}-{external_id}-state"
-            df_to_be_written = None
-            if self.state_store.get_state(state_id)[0] is None:
-                df_to_be_written = df
-            else:
-                latest_written_time = self.state_store.get_state(state_id)[0]
-                df_to_be_written = df[df["timestamp"] > latest_written_time]
-
-            if len(df_to_be_written) > 0:
-                latest_process_time = df_to_be_written["timestamp"].max()
-
-                df_to_be_written.loc[:, "externalId"] = df_to_be_written.apply(
-                    lambda row: self.config.destination.time_series_prefix
-                    + row["externalId"],
-                    axis=1,
-                )
-                df_to_be_written = df_to_be_written.pivot(
-                    index="timestamp", columns="externalId", values="value"
-                )
-                df_to_be_written.index = pd.to_datetime(df_to_be_written.index)
-                try:
-                    self.client.time_series.data.insert_dataframe(df_to_be_written)
-                except CogniteNotFoundError as notFound:
-                    missing_xids = notFound.not_found
-                    for new_timeserie in missing_xids:
-                        # Create the missing time series
-                        self.client.time_series.create(
-                            TimeSeriesWrite(
-                                name=new_timeserie["externalId"],
-                                external_id=new_timeserie["externalId"],
-                                is_string=True,
-                                data_set_id=self.data_set_id,
-                            )
-                        )
-                except CogniteAPIError as e:
-                    self.logger.error(
-                        f"Error while writing time series data to CDF: {e}"
+            latest_timestamps[external_id] = self.state_store.get_state(state_id)[0]
```

Comment:
get_state can take a list of external ids as an argument, we should use that to avoid too many roundtrips if you have many external ids in your table
https://cognite-extractor-utils.readthedocs-hosted.com/en/latest/docs.html#cognite.extractorutils.statestore.AbstractStateStore.get_state

Reply:
Thanks for the suggestion, modified code to use this